---
date: "`r Sys.Date()`"
output: html_document
link-citations: true
params:
    primary_genome: ""
    compare_genome: ""
title: "Identifying `r params$primary_genome` contigs using `r params$compare_genome` as a reference"
---

```{r file_info, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, dev='png') 

#To re-run this analysis with different assemblies, input the names of the desired assemblies below

genome1 <- params$primary_genome
genome2 <- params$compare_genome

#Include an accurate prefix location for results files

prefix <- "results/"
```


```{r Packages, message=FALSE}
if (!requireNamespace("slanter", quietly = TRUE)) {
  install.packages("slanter", repos='[https://cloud.r-project.org](https://cloud.r-project.org/)')
}

library(tidyverse)
library(kableExtra) # Formatting Tables
library(slanter) # Plotting Heatmap
library(ggrepel) # Visualizing homologus scaffold sizes
```

```{r, echo=FALSE}
quast_1 <- read.csv(paste0(prefix,genome1,"/",genome1,"_quast_table.txt"), header=FALSE)
quast_2 <- read.csv(paste0(prefix,genome2,"/",genome2,"_quast_table.txt"), header=FALSE)

busco_1 <- read.csv(paste0(prefix,genome1,"/",genome1,"_busco_table.txt"), header=FALSE)
busco_2 <- read.csv(paste0(prefix,genome2,"/",genome2,"_busco_table.txt"), header=FALSE)
```


```{r Tables, echo=FALSE}
quast_table <- tribble(
   ~`QUAST Statistics`, ~`Contigs`, ~`L50`, ~`L75`, ~`Total Length`,
   quast_1$V1[1], quast_1$V1[2], quast_1$V1[3], quast_1$V1[4], paste0(round(as.numeric(quast_1$V1[5])/1000000)," mb"),
   quast_2$V1[1], quast_2$V1[2], quast_2$V1[3], quast_2$V1[4], paste0(round(as.numeric(quast_2$V1[5])/1000000)," mb"),
)

busco_table <- tribble(
   ~`% of 5286 Conserved Genes`, ~`Complete`, ~`Duplicated `, ~`Missing`,
   busco_1$V1[1], busco_1$V1[2], busco_1$V1[3], busco_1$V1[4],
   busco_2$V1[1], busco_2$V1[2], busco_2$V1[3], busco_2$V1[4],
)

quast_table %>%
  kable("html", align = 'clc') %>%
    kable_styling(full_width = F, position = "float_left")
 
busco_table %>%
  kable("html", align = 'clc') %>%
    kable_styling(full_width = F, position = "right")
```

### Scaffold Overview

```{r}
scaffolds_1 <- read.csv(paste0(prefix,genome1,"/",genome1,".fa.fai"),sep="\t", header=FALSE) %>% select(V1, V2)
scaffolds_2 <- read.csv(paste0(prefix,genome2,"/",genome2,".fa.fai"),sep="\t", header=FALSE) %>% select(V1, V2)


scaffolds_1$assembly <- genome1
scaffolds_2$assembly <- genome2
combined_scaffold_table <- rbind(scaffolds_1,scaffolds_2)
```

```{r, fig.height=4, fig.width=8}
ggplot(combined_scaffold_table, aes(x = V2, fill=assembly)) +
 geom_histogram( color="#e9ecef", alpha=0.6, bins = 50) +
    scale_fill_manual(values=c("#69b3a2", "#404080")) +
    scale_x_log10() +
  scale_y_log10() +
    theme_bw() +
  ylab("Count") +
  xlab("Contig Size")+
  theme(text=element_text(size=18))
```

### BUSCO Overview

Here we load data generated by BUSCO, which contains BUSCO IDs of conserved genes along with scaffold and base pair locations of the gene in the assembly. See below a summary visual of BUSCO gene number per scaffold. Empty category names correspond to BUSCO genes not found anywhere in the assemblies.

```{r BUSCO Data, figures-side, fig.show="hold", out.width="50%"}
#These tables must have the headers fixed with underscores

### Open genome1 BUSCO results
full_busco_1 <- read.csv(paste0(prefix,genome1,"/",genome1,"_full_busco_table.txt"),sep="\t", header=TRUE) %>% mutate(Sequence = str_remove(Sequence, ":.*")) 

# Group results by scaffold name, and count genes per scaffold
# Empty row name corresponds to BUSCO genes that were not found in the assembly
full_busco_table_1 <- full_busco_1 %>% 
 group_by(Sequence) %>% summarise(n = n()) %>% arrange(n) %>% as_tibble()

# Plot number of BUSCO genes per scaffold
ggplot(full_busco_table_1, aes(x=reorder(Sequence, -n), y=n)) +
  geom_bar(stat = "identity") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.title.x = element_text(size=18), axis.title.y = element_text(size = 18)) +  
  xlab(paste0(genome1," Contig Name")) +
  ylab("Number of BUSCO Genes")

### Open genome2 BUSCO results
full_busco_2 <- read.csv(paste0(prefix,genome2,"/",genome2,"_full_busco_table.txt"),sep="\t", header=TRUE) %>%  mutate(Sequence = str_remove(Sequence, ":.*"))

# Group results by scaffold name, and count genes per scaffold
# Empty row name corresponds to BUSCO genes that were not found in the assembly
full_busco_table_2 <- full_busco_2 %>%
 group_by(Sequence) %>% summarise(n = n()) %>% arrange(n) %>% as_tibble()

# Plot number of BUSCO genes per scaffold
ggplot(full_busco_table_2, aes(x=reorder(Sequence, -n), y=n)) +
  geom_bar(stat = "identity") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.title.x = element_text(size=18), axis.title.y = element_text(size = 18))+ 
  xlab(paste0(genome2," Contig Name")) +
  ylab("Number of BUSCO Genes") 

```

Next we extract lists of BUSCO genes present on each scaffold, then make every possible pairwise overlap comparison between our two genomes' scaffolds. Overlaps of at least one gene but fewer than 8 genes represent small deviations between assemblies, including some genes that are moved, duplicated, or found in only one assembly. These are stored for further analysis. Overlaps of 5 or more genes are used to assign homology between scaffolds in the two assemblies. A list of homologous assemblies is printed below, accompanied by a heatmap visual of gene overlap.

```{r Gene Overlap Matrix, fig.height=8, fig.width=8}
# Extract the BUSCO gene names per scaffold for genome1
full_busco_1$Sequence <- factor(full_busco_1$Sequence) 
full_busco_processed_1 <- full_busco_1 %>% group_by(full_busco_1$Sequence) %>% 
                                      summarize(TheList = list(Busco_id)) %>% select(TheList)

busco_list_1 <- full_busco_processed_1[[1]]
names(busco_list_1) <- levels(full_busco_1$Sequence)


# Extract the BUSCO gene names per scaffold for genome2
full_busco_2$Sequence <- factor(full_busco_2$Sequence) 
full_busco_processed_2 <-full_busco_2 %>%
 group_by(full_busco_2$Sequence) %>% summarize(TheList = list(Busco_id)) %>% select(TheList)

busco_list_2 <- full_busco_processed_2[[1]]
names(busco_list_2) <- levels(full_busco_2$Sequence)


# Build an empty matrix to hold all pairwise overlap comparisons
x  <- matrix(NA_real_, length(busco_list_1), length(busco_list_2))
rownames(x) <- names(busco_list_1) 
colnames(x) <- names(busco_list_2)

found.or.moved.genes <- c() # Makes list of genes present in small overlaps
scaffold.pairs <- c() # Stores a list of scaffolds with more than 5 overlapping BUSCOs

# Loop through all pairs of hmel2.5 x hmr scaffolds and quantify busco gene overlap
for (i in seq_along(busco_list_1)){
  for (j in seq_along(busco_list_2)){ 
    
      # Creates a list of overlapped genes
      gene.list <- intersect(unlist(busco_list_1[i]), unlist(busco_list_2[j]))
      
      # Creates a list of genes that represent small, unexpected overlaps (genes only in one assembly or assembled onto different scaffolds)
      if (length(gene.list) > 0 && length(gene.list) < 5) { found.or.moved.genes <- c(found.or.moved.genes,gene.list)
      #cat(names(busco_list_2[j]), "\t", names(busco_list_1[i]), "\t", length(gene.list), "\n")
      }
      
      # Creates and print out a list of scaffold pairs with 5 or more BUSCO gene overlaps
      if (length(gene.list) > 4) { 
        
          current.pair <- cbind(names(busco_list_2[j]), names(busco_list_1[i]), length(gene.list))
          scaffold.pairs <- rbind(scaffold.pairs,current.pair)
      
        }
      x[i,j] <- length(gene.list)
    }
}

scaffold.pairs <- as_tibble(scaffold.pairs)
kable(scaffold.pairs)

# Creates a modified overlap matrix and plots the result
x_adj = x
x_adj = x_adj + 1 #  Changed to better visualize small non-zero overlap numbers, and to prevent NA when taking log2
slanter::sheatmap(log2(x_adj),cluster_rows = FALSE,cluster_cols = FALSE)


```
  
Genome 1 is the y axis, genome 2 is the x axis. The data matrix has its values increased by one (to eliminate 0 values) and then log2 transformed to facilitate visualization of small BUSCO overlap values.  

Finally, we view the relative sizes, in basepairs, of scaffolds identified as homologus.

```{r}
scaffold.lengths <- left_join(scaffold.pairs, scaffolds_2, by = c("V1")) %>% left_join(scaffolds_1, by = c("V2.x" = "V1")) %>% select("V1", "V2.y", "V2.x", "V2", "V3")

length2 <- paste0(genome2,"_length")
length1 <- paste0(genome1,"_length")

#names(scaffold.lengths) <- c(genome2, length2, genome1, length1, "overlaps")

scaffold.lengths
```



```{r}
ggplot(scaffold.lengths, aes(x=V2.y, y=V2)) +
      geom_point() +
      theme(legend.position="none") +
      xlab(paste0(genome2,"_length")) + 
      ylab(length1 <- paste0(genome1,"_length")) +
      geom_label_repel(aes(label = V2.x),
                  box.padding   = 0.35, 
                  point.padding = 0.5,
                  segment.color = 'grey50')


```




```{r Session Information}
sessionInfo()
```