---
title: "ATAC-Seq"
author: "Daniel and Krista"
date: "2024-09-26"
output: html_document
bibliography: ATAC-Seq-Citation-Cortex.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this RMD, you will find ATAC-Seq data overlayed with our coverage and SNP frequency plots at the Cortex locus. The ATAC-Seq data comes from @Livraghi2021. First, I logged in to DCC (ssh [dsl42\@dcc-login.oit.duke.edu](mailto:dsl42@dcc-login.oit.duke.edu){.email}). Then, becase we know the ATAC-seq data is stored in the SRA database...

ssh [dsl42\@dcc-login.oit.duke.edu](mailto:dsl42@dcc-login.oit.duke.edu){.email} module load SRA-Toolkit/3.0.0-rhel8 fasterq-dump ERR5466846 - this is a tool SRA database made. we use it to download data from database! - no path needed or any baggage. jsut give it the file or all 100 files and it will download RELEVANT reads (forward and reverse, for ex) fasta - sequence of every read bam - sequence + location - get from fasta to bam through bwa aligner

slurm - when a super computer has a lot of users, slurm is the program that decides who's requests run first. it decides who gets ram and for how long. - we are asking for a big taks - search a gigabite of sequences - millions - against a .3 gigs genome - larger than a novel- - we are very respectful - well tell it hey this is big, FYI get ready and allocate RAM as you see fit :) - this is slurm

now we are going to talk to slurm by making bash file and putting request at the top

vim alignment_for_cortex.sh click I then past #!/bin/bash #SBATCH --mem=100G #SBATCH --cpus-per-task 24 #SBATCH --partition scavenger - this code: 1st line - im bash so run me iwth bash 2 - give me 100 gb of ram, 3 - give me 24 cpu cors 4 - telling slurm the type of resoruces we are asking for - we are asking for scavengar - if labs that arnt using ram and cpu that they bought lets use it

now get out of there by doing esc :wq

module avail module load BWA/0.7.17 bwa index Heliconius_melpomene_melpomene_Hmel2.5.scaffolds.fa - the fasta is the book (.fa), and .fai = making an index at front of the book

vim alignment_for_cortex.sh click I bwa mem Heliconius_melpomene_melpomene_Hmel2.5.scaffolds.fa ERR5466846_1.fastq ERR5466846_2.fastq \> ERR5466846_aligned.sam esc :wq

now submit the bash script (we made with vim - the conduit) to slurm to manage sbatch alignment_for_cortex.sh

check on how it is doing with your job: squeue -u dsl42

if it does not work, ls then see what new file it gave you, then vim the new file and see the message

now lets do for the second atacseq file that we already put in cp alignment_for_cortex.sh 2nd_alignmnet_for_cortex.sh - tells it to take the code from the first vim that we did and put it into another file - this will be the file for the second atac seq data we are using. vim 2nd_alignmnet_for_cortex.sh - and inside here, update the ERRs

we then installed miniconda ("python" for science) by following directions on this website: <https://oit-rc.pages.oit.duke.edu/rcsupportdocs/software/user/>

COOL. then close out so it can take a rest.

now we are having pip install packages for us so that it can do the work and we don't have to pip install MACS2 - this is the software we are using for peak calling

check that it works macs3 callpeak -h

now turn sam into bam module load samtools/1.9 
samtools view -bS ERR5466846_aligned.sam \> ERR5466846_aligned.bam

(DO NOT DO THESE LAST 5 THINGS FOR COVERAGE)
now, prep data to give to slurm vim peak_call_1st_atac.sh inside, #!/bin/bash #SBATCH --mem=100G #SBATCH --cpus-per-task 24 #SBATCH --partition scavenger macs3 callpeak -t ERR5466846_aligned.bam -n ERR5466846 -f BAMPE #the extra stuff: -t tells it look at this file, -n gives it a name for output file.

esc :wq

sbatch peak_call_1st_atac.sh - gives aligned bam and atac to slurm and tells slurm to call peaks

check that slurm is working iwth squeue -u dsl42

now pull it onto your computer using ubuntu (open a new ubuntu window, not just + new tab) scp [dsl42\@dcc-login.oit.duke.edu](mailto:dsl42@dcc-login.oit.duke.edu){.email}:/hpc/home/dsl42/ERR5466846_peaks.narrowPeak /mnt/c/Users/danie/OneDrive/Documents/Wray_Lab/gynandromorph_2024/

10/1 1. login to dcc

(pickup here for coverage)
ssh [dsl42\@dcc-login.oit.duke.edu](mailto:dsl42@dcc-login.oit.duke.edu){.email} module load samtools/1.9 - sorts the bam

samtools sort ERR5466846_aligned.bam -o ERR5466846_sorted.bam

-   imput the bam you want sorted (by genome position) and then the output file name you want to see

module load Bedtools/2.30.0 bedtools genomecov -bga -ibam ERR5466846_sorted.bam \> ERR5466846.bedgraph

-   we are using -d so that we have a coverage (from atacseq) value at every position for R vim ERR5466846.bedgraph

we got the bedtools code from <https://bedtools.readthedocs.io/en/latest/content/tools/genomecov.html>

now save the ERR5466846.bedgraph to your computer by pulling it from super computer to yours

write this in ubuntu (open a new ubuntu window, not just + new tab) scp [dsl42\@dcc-login.oit.duke.edu](mailto:dsl42@dcc-login.oit.duke.edu){.email}:/hpc/home/dsl42/ERR5466846.bedgraph /mnt/c/Users/danie/OneDrive/Documents/Wray_Lab/gynandromorph_2024/

now we need to free up space so moving things

idiogram heatmap of how simple and complex are distributed - idogram on non-model organism - hard - centromeric vs holocentric distribution of repetitive elements? - is it repetitive sequence that accounts for size dif in mel vs erato and if so what kind a one area? are all chromosomes bigger? - is there lepidopetera that is not locoentric

!!! if you stored file in /work/dsl42 take out /hpc/ from the code you use to pull it onto your computer 

::: {#refs}
:::

```{r Session Information}
  sessionInfo()

```
