

rule targets:
    input:
        "../results/bc2011.linear_plot.png", #fasta_qc
        "../results/allAnnotation_bc2011.gff" #annotations
        

rule assembly:
    input:
        rawDataPath = "../raw_data/TN2L_S1.fasta"                                    
    output:
        "../analysis/{sample}.p_ctg.fa",
        "../logs/{sample}.log"
    shell:
        """
        bash hifiAssembly.sh {wildcards.sample} {input.rawDataPath}
        """

rule fasta_qc:
    input:
        rawDataPath = "../raw_data/TN2L_S1.fasta"                                    

    output:
        "../results/{sample}.linear_plot.png"   #can you make genomescope output to add name** bash command #"fastaAnalyses/contig_{sample}.txt"
    shell:
        """
        bash kmerAnalysis.sh {input.rawDataPath}
        """


rule masking:
    input: 
        "../analysis/{sample}.p_ctg.fa"
    output:
        "../results/{sample}.p_ctg.masked.fasta"
    shell:
        """
        bash scripts/maskingFasta.py"
        """

rule geneAnnotations:
    input:
        "../results/{sample}.p_ctg.masked.fasta"
    output:
        "../analysis/geneAnnotation_{sample}.gff"
    script:
        "scripts/geneGFF.py" 

rule noncodingAnnotations:
    input:
        "../results/{sample}.p_ctg.masked.fasta"
    output:
        "../analysis/noncodingAnnotation_{sample}.gff"
    script:
        "scripts/noncodingGFF.py" 

rule combinedGFF:
    input:
        "../analysis/geneAnnotation_{sample}.gff",
        "../analysis/noncodingAnnotation_{sample}.gff"
    output:
        "../results/allAnnotation_{sample}.gff"
    script:
        "combineGFF.py"